src/Lean/Linter/Basic.lean

```diff
@@ -0,0 +1,130 @@
+import Lean.Elab.Command
+import Lean.Linter.Util
+import Lean.Elab.InfoTree
+import Lean.Server.InfoUtils
+import Lean.Server.References
+import Std.Data.HashMap
+
+namespace Lean.Linter
+open Lean.Server Std
+
+register_builtin_option linter.unusedVariables : Bool := {
+  defValue := true,
+  descr := "enable the 'unused variables' linter"
+}
+
+def getLinterUnusedVariables (o : Options) : Bool := o.get linter.unusedVariables.name (getLinterAll o)
+
+def unusedVariables : Linter := fun stx => do
+  let some stxRange := stx.getRange?
+    | pure ()
+
+  let infoTrees := (← get).infoState.trees.toArray
+  let fileMap := (← read).fileMap
+
+  -- collect references
+  let refs := findModuleRefs fileMap infoTrees
+
+  let mut vars : HashMap FVarId RefInfo := .empty
+  let mut constDecls : HashSet String.Range := .empty
+
+  for (ident, info) in refs.toList do
+    match ident with
+    | .fvar id =>
+      vars := vars.insert id info
+    | .const _ =>
+      if let some definition := info.definition then
+        if let some range := definition.stx.getRange? then
+          constDecls := constDecls.insert range
+
+  -- collect uses from tactic infos
+  let tacticMVarAssignments : HashMap MVarId Expr :=
+    infoTrees.foldr (init := .empty) fun tree assignments =>
+      tree.foldInfo (init := assignments) (fun _ i assignments => match i with
+        | .ofTacticInfo ti =>
+          ti.mctxAfter.eAssignment.foldl (init := assignments) fun assignments mvar expr =>
+            if assignments.contains mvar then
+              assignments
+            else
+              assignments.insert mvar expr
+        | _ =>
+          assignments)
+
+  let tacticFVarUses : HashSet FVarId ←
+    tacticMVarAssignments.foldM (init := .empty) fun uses _ expr => do
+      let (_, s) ← StateT.run (s := uses) <| expr.forEach' (fun expr => do
+        match expr with
```

Comment:
Note that this will not visit assignments of nested metavariables, so we could miss a part of the proof. But since we do this for every tactic step, I don't think this is an issue for the current set of tactics.

Reply:
(nevermind, you're iterating over `eAssignments` directly, not `goalsAfter`)